From 2675823ee87ce12afb06b9abf800e5a5bcecd1b4 Mon Sep 17 00:00:00 2001
From: Liwei Song <liwei.song@windriver.com>
Date: Fri, 21 Feb 2020 11:23:19 +0800
Subject: [PATCH] Allow sshd_t and staff_t read symlinks

when enable ostree, root directory is symlink, sshd_t and staff_t
do not have permission to access link file, add rules to allow
this kind of access.
add rules when root is label or dircetory to userdomain, alias in
file_contexts.subs_dist for root will stop label for /root if it
is symlink file.

fix the following denials:
AVC avc:  denied  { read } for  pid=733 comm="sshd" name="root"
dev="sdb3" ino=861 scontext=system_u:system_r:sshd_t:s0-s15:c0.c1023
tcontext=root:object_r:user_home_dir_t:s0 tclass=lnk_file permissive=0

AVC avc:  denied  { read } for  pid=733 comm="sh" name="root"
dev="sdb3" ino=861 scontext=root:staff_r:staff_t:s0-s15:c0.c1023
tcontext=root:object_r:user_home_dir_t:s0 tclass=lnk_file permissive=0

Upstream-Status: Inappropriate [ostree specific]

Signed-off-by: Liwei Song <liwei.song@windriver.com>
---
 config/file_contexts.subs_dist      | 4 ----
 policy/modules/roles/staff.te       | 1 +
 policy/modules/services/ssh.te      | 2 ++
 policy/modules/system/userdomain.fc | 3 +++
 4 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/config/file_contexts.subs_dist b/config/file_contexts.subs_dist
index 9ba16c626..60a2f0506 100644
--- a/config/file_contexts.subs_dist
+++ b/config/file_contexts.subs_dist
@@ -46,10 +46,6 @@
 /usr/lib/busybox/sbin /usr/sbin
 /usr/lib/busybox/usr /usr
 
-# The genhomedircon.py will expand /root home directory to /home/root
-# Add an aliase for it
-/root /home/root
-
 /usr/etc /etc
 /var/home /home
 /var/rootdirs/root /root
diff --git a/policy/modules/roles/staff.te b/policy/modules/roles/staff.te
index af069f0ad..51f03ef8c 100644
--- a/policy/modules/roles/staff.te
+++ b/policy/modules/roles/staff.te
@@ -14,6 +14,7 @@ userdom_unpriv_user_template(staff)
 # Local policy
 #
 corenet_ib_access_unlabeled_pkeys(staff_t)
+files_read_all_symlinks(staff_t)
 
 optional_policy(`
 	apache_role(staff, staff_t, staff_application_exec_domain, staff_r)
diff --git a/policy/modules/services/ssh.te b/policy/modules/services/ssh.te
index a93f2447d..dadc1eca0 100644
--- a/policy/modules/services/ssh.te
+++ b/policy/modules/services/ssh.te
@@ -73,6 +73,8 @@ userdom_user_tmpfs_file(ssh_tmpfs_t)
 type ssh_home_t;
 userdom_user_home_content(ssh_home_t)
 
+files_read_all_symlinks(sshd_t)
+
 type sshd_keytab_t;
 files_type(sshd_keytab_t)
 
diff --git a/policy/modules/system/userdomain.fc b/policy/modules/system/userdomain.fc
index 173e314af..83c6dd80d 100644
--- a/policy/modules/system/userdomain.fc
+++ b/policy/modules/system/userdomain.fc
@@ -4,6 +4,9 @@ HOME_DIR/bin(/.*)?		gen_context(system_u:object_r:user_bin_t,s0)
 HOME_DIR/\.local/bin(/.*)?		gen_context(system_u:object_r:user_bin_t,s0)
 HOME_DIR/\.pki(/.*)?	gen_context(system_u:object_r:user_cert_t,s0)
 
+/root		-d	gen_context(root:object_r:user_home_dir_t,s0)
+/root		-l	gen_context(root:object_r:user_home_dir_t,s0)
+
 /tmp/gconfd-%{USERNAME} -d	gen_context(system_u:object_r:user_tmp_t,s0)
 
 /run/user		-d	gen_context(system_u:object_r:user_runtime_root_t,s0)
-- 
2.25.1

